
<h1 align="center"> Battle of the Neighbourhoods - Toronto </h1>

Author: Aman Attar

This notebook contains Questions 1, 2 & 3 of the Assignment. They have been segregated by Section headers


In [1]:
import pandas as pd

# Question 1

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]

In [8]:
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [10]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [11]:
df = df.reset_index()

df

In [13]:
df.drop(['index'], axis = 'columns', inplace= True)

In [14]:
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [15]:
df.shape

(103, 3)

Answer to Question 1: We have 103 rows and 3 columns

# Question 2

In [16]:
#pip install gecoder

In [17]:
import geocoder #geocodeer is imorted

Tried below approach but the execution doesn't stop

```python
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
```

In [22]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [23]:
print("The Shape of our wiki data is : ",df.shape)
print("The shape of csv data is : ",data.shape)

The Shape of our wiki data is :  (103, 3)
The shape of csv data is :  (103, 3)


In [25]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [26]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [27]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how = 'inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [28]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

# Question 3

Drawing inspiration from the previous lab where we cluster the neighbourhood of NYC, We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [30]:
from geopy.geocoders import Nominatim

In [31]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [32]:
import folium

In [34]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
 
map_Toronto

In [35]:
CLIENT_ID = 'IN41YKELDAZOEUVSKZNKIGO4BYD43Z4DROSSVZI4DBO0ZO4Q' 
CLIENT_SECRET = 'CKYMYQASJ0OSLNB2OVOFPQIQXJNFK2VPUEBRRVMEVGVE511Q'
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: IN41YKELDAZOEUVSKZNKIGO4BYD43Z4DROSSVZI4DBO0ZO4Q
CLIENT_SECRET:CKYMYQASJ0OSLNB2OVOFPQIQXJNFK2VPUEBRRVMEVGVE511Q


Next, we create a function to get all the venue categories in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [37]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [39]:
venues_in_toronto.shape

(1325, 5)

In [40]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                               Venue              Venue Category  
0                    Brookbanks Park                        Park  
1                      Variety Store           Food & Drink Shop  
2  Corrosion Service Company Limited  Construction & Landscaping  
3             Victoria Village Arena                Hockey Arena  
4                          Portugril       Portuguese Restaurant

In [41]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1311  Mimico NW, The Queensway West, South of Bloor,...   
1312  Mimico NW, The Queensway West, South of Bloor,...   
1313  Mimico NW, The Queensway West, South of Bloor,...   
1314  Mimico NW, The Queensway West, South of Bloor,...   
1315  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude  \
0                  43.753259               -79.329656   
1                  43.753259               -79.329656   
2                  43.753259               -79.329656   
3                  43.725882               -79.315572   
4                  43.725882               -79.315572   
...                      ...                      ...   
1311               43.628841               -79.520999   
1312               43.628841               -79.520999   
1313               43.628841               -79.520999   
1314               43.628841               -79.520999   
1315               43.628841               -79.520999   

                                  Venue              Venue Category  
0                       Brookbanks Park                        Park  
1                         Variety Store           Food & Drink Shop  
2     Corrosion Service Company Limited  Construction & Landscaping  
3                Victoria Village Arena                Hockey Arena  
4                             Portugril       Portuguese Restaurant  
...                                 ...                         ...  
1311                         Wingporium                 Wings Joint  
1312                   South St. Burger                Burger Joint  
1313                          Dollarama              Discount Store  
1314                     Healthy Planet             Supplement Shop  
1315               Artisano Bakery Café                      Bakery  

[411 rows x 5 columns]

In [42]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                                   Wexford, Maryvale   
Adult Boutique                                   Church and Wellesley   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.750072               -79.295849   
Adult Boutique                   43.665860               -79.383160   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.653206               -79.379817   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.340923   

                                              Venue  
Venue Category                                       
Accessories Store                       Puffin Gear  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                          The National Club  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[235 rows x 4 columns]

In [43]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1320                  0               0        0                   0   
1321                  0               0        0                   0   
1322                  0               0        0                   0   
1323                  0               0        0                   0   
1324                  0               0        0                   0   

      Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0                0               0                0                 0   
1                0               0                0                 0   
2                0               0                0                 0   
3                0               0                0                 0   
4                0               0                0                 0   
...            ...             ...              ...               ...   
1320             0               0                0                 0   
1321             0               0                0                 0   
1322             0               0                0                 0   
1323             0               0                0                 0   
1324             0               0                0                 0   

      American Restaurant  Antique Shop  ...  Truck Stop  \
0                       0             0  ...           0   
1                       0             0  ...           0   
2                       0             0  ...           0   
3                       0             0  ...           0   
4                       0             0  ...           0   
...                   ...           ...  ...         ...   
1320                    0             0  ...           0   
1321                    0             0  ...           0   
1322                    0             0  ...           0   
1323                    0             0  ...           0   
1324                    0             0  ...           0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1320                              0                 0            0   
1321                              0                 0            0   
1322                              0                 0            0   
1323                              0                 0            0   
1324                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1320                      0                0         0            0   
1321                      0                0         0            0   
1322                      0                0         0         

In [44]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  Airport  \
0         Parkwoods                  0               0        0   
1         Parkwoods                  0               0        0   
2         Parkwoods                  0               0        0   
3  Victoria Village                  0               0        0   
4  Victoria Village                  0               0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  ...  Truck Stop  \
0                 0                    0  ...           0   
1                 0                    0  ...           0   
2                 0                    0  ...           0   
3                 0                    0  ...           0   
4                 0                    0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 236 columns]

In [45]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0             0.0      0.0                 0.0           0.0             0.0   
1             0.0      0.0                 0.0           0.0             0.0   
2             0.0      0.0                 0.0           0.0             0.0   
3             0.0      0.0                 0.0           0.0             0.0   
4             0.0      0.0                 0.0           0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  ...  Truck Stop  \
0              0.0               0.0             0.000000  ...         0.0   
1              0.0               0.0             0.000000  ...         0.0   
2              0.0               0.0             0.000000  ...         0.0   
3              0.0               0.0             0.000000  ...         0.0   
4              0.0               0.0             0.041667  ...         0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 236 columns]

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [47]:
import numpy as np

In [48]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Breakfast Spot          Skating Rink   
1                        Gym          Skating Rink           Coffee Shop   
2                       Bank   Fried Chicken Joint        Ice Cream Shop   
3                       Bank   Japanese Restaurant                  Café   
4                Pizza Place           Coffee Shop        Sandwich Place   

  5th Most Common Venue      6th Most Common Venue  \
0  Ethiopian Restaurant          Electronics Store   
1                   Pub                   Pharmacy   
2           Supermarket  Middle Eastern Restaurant   
3           Yoga Studio               Dessert Shop   
4          Liquor Store                 Restaurant   

         7th Most Common Venue  8th Most Common Venue  \
0  Eastern European Restaurant              Drugstore   
1               Airport Lounge  College Arts Building   
2                 Intersection             Restaurant   
3         Ethiopian Restaurant      Electronics Store   
4             Sushi Restaurant                Butcher   

         9th Most Common Venue       10th Most Common Venue  
0                   Donut Shop                      Dog Run  
1            Electronics Store  Eastern European Restaurant  
2                Deli / Bodega                  Pizza Place  
3  Eastern European Restaurant                    Drugstore  
4                          Pub                         Café

In [49]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [50]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [51]:
kmeans.labels_[0:100]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 3, 0, 1, 1, 1, 0, 1, 1, 1, 2, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1,
       0, 1, 0, 1, 1, 1, 1, 0])

In [52]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [53]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             0.0                  Park   
1  43.725882 -79.315572             1.0           Coffee Shop   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

        2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0  Construction & Landscaping      Food & Drink Shop           Yoga Studio   
1           French Restaurant  Portuguese Restaurant          Hockey Arena   
2                        Park                 Bakery        Breakfast Spot   
3      Furniture / Home Store      Accessories Store              Boutique   
4            Sushi Restaurant            Yoga Studio   Fried Chicken Joint   

         5th Most Common Venue  6th Most Common Venue  \
0                Deli / Bodega      Electronics Store   
1  Eastern European Restaurant              Drugstore   
2                      Theater      French Restaurant   
3                  Event Space     Miscellaneous Shop   
4               Sandwich Place  Portuguese Restaurant   

         7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0  Eastern European Restaurant             Drugstore             Donut Shop   
1                   Donut Shop               Dog Run    Distribution Center   
2          Distribution Center                   Spa            Event Space   
3                 Carpet Store           Coffee Shop  Vietnamese Restaurant   
4                         Park             Nightclub     Mexican Restaurant   

  10th Most Common Venue  
0                Dog Run  
1         Discount Store  
2           Dessert Shop  
3        Airport Service  
4    Japanese Restaurant

In [54]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [55]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [56]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Cluster 1

In [57]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             0.0                  Park   
21              York             0.0                  Park   
32       Scarborough             0.0            Playground   
35         East York             0.0                  Park   
49        North York             0.0                  Park   
52        North York             0.0                  Park   
61   Central Toronto             0.0                  Park   
64              York             0.0                  Park   
66        North York             0.0                  Park   
68   Central Toronto             0.0                  Park   
77         Etobicoke             0.0                  Park   
85       Scarborough             0.0                  Park   
91  Downtown Toronto             0.0                  Park   

         2nd Most Common Venue       3rd Most Common Venue  \
0   Construction & Landscaping           Food & Drink Shop   
21               Women's Store                        Pool   
32                 Yoga Studio                Dance Studio   
35           Convenience Store                Intersection   
49            Basketball Court  Construction & Landscaping   
52                 Yoga Studio                Dance Studio   
61                 Swim School                    Bus Line   
64           Convenience Store               Jewelry Store   
66  Construction & Landscaping           Convenience Store   
68               Jewelry Store            Sushi Restaurant   
77                    Bus Line              Sandwich Place   
85                  Playground                Intersection   
91                  Playground                       Trail   

   4th Most Common Venue        5th Most Common Venue  \
0            Yoga Studio                Deli / Bodega   
21           Yoga Studio                 Dance Studio   
32     Electronics Store  Eastern European Restaurant   
35           Yoga Studio                Deli / Bodega   
49                Bakery                  Yoga Studio   
52     Electronics Store  Eastern European Restaurant   
61           Yoga Studio                Deli / Bodega   
64           Yoga Studio                Deli / Bodega   
66           Yoga Studio                Deli / Bodega   
68                 Trail                 Dance Studio   
77     Mobile Phone Shop      Comfort Food Restaurant   
85           Yoga Studio                 Dance Studio   
91           Curling Ice  Eastern European Restaurant   

          6th Most Common Venue        7th Most Common Venue  \
0             Electronics Store  Eastern European Restaurant   
21  Eastern European Restaurant                    Drugstore   
32                    Drugstore                   Donut Shop   
35            Electronics Store  Eastern European Restaurant   
49                 Dessert Shop         Ethiopian Restaurant   
52                    Drugstore                   Donut Shop   
61  Eastern European Restaurant                    Drugstore   
64            Electronics Store  Eastern European Restaurant   
66            Electronics Store  Eastern European Restaurant   
68  Eastern European Restaurant                    Drugstore   
77                   Comic Shop  Eastern European Restaurant   
85  Eastern European Restaurant                    Drugstore   
91                    Drugstore                   Donut Shop   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0              Drugstore                   Donut Shop                Dog Run  
21            Donut Shop                      Dog Run    Distribution Center  
32               Dog Run          Distribution Center         Discount Store  
35             Drugstore                   Donut Shop                Dog Run  
49     Electronics Store  Eastern European Restaurant              Drugstore  
52               Dog Run          Distribution Center         Discount Store  
61            Donut Shop  

Cluster 2

In [58]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             1.0           Coffee Shop   
2    Downtown Toronto             1.0           Coffee Shop   
3          North York             1.0        Clothing Store   
4    Downtown Toronto             1.0           Coffee Shop   
7          North York             1.0                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto             1.0                  Café   
98          Etobicoke             1.0                  Pool   
99   Downtown Toronto             1.0           Coffee Shop   
100      East Toronto             1.0                  Park   
102         Etobicoke             1.0                   Gym   

      2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
1         French Restaurant  Portuguese Restaurant          Hockey Arena   
2                      Park                 Bakery        Breakfast Spot   
3    Furniture / Home Store      Accessories Store              Boutique   
4          Sushi Restaurant            Yoga Studio   Fried Chicken Joint   
7       Japanese Restaurant            Coffee Shop            Restaurant   
..                      ...                    ...                   ...   
97              Coffee Shop             Restaurant    Seafood Restaurant   
98                    River            Yoga Studio           Curling Ice   
99         Sushi Restaurant                   Park          Burger Joint   
100           Garden Center          Auto Workshop         Burrito Place   
102          Hardware Store            Flower Shop       Supplement Shop   

           5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
1    Eastern European Restaurant              Drugstore            Donut Shop   
2                        Theater      French Restaurant   Distribution Center   
3                    Event Space     Miscellaneous Shop          Carpet Store   
4                 Sandwich Place  Portuguese Restaurant                  Park   
7                     Beer Store            Supermarket    Italian Restaurant   
..                           ...                    ...                   ...   
97                         Hotel            Art Gallery          Concert Hall   
98   Eastern European Restaurant              Drugstore            Donut Shop   
99             Indian Restaurant   Ethiopian Restaurant              Beer Bar   
100         Fast Food Restaurant         Farmers Market                Garden   
102                       Bakery           Burger Joint        Discount Store   

    8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
1                 Dog Run    Distribution Center         Discount Store  
2                     Spa            Event Space           Dessert Shop  
3             Coffee Shop  Vietnamese Restaurant        Airport Service  
4               Nightclub     Mexican Restaurant    Japanese Restaurant  
7          Discount Store     Dim Sum Restaurant     Chinese Restaurant  
..                    ...                    ...                    ...  
97            Pizza Place                    Pub         Sandwich Place  
98                Dog Run    Distribution Center         Discount Store  
99    Japanese Restaurant              Bookstore         Breakfast Spot  
100            Restaurant             Comic Shop            Pizza Place  
102     Convenience Store         Sandwich Place   Fast Food Restaurant  

[83 rows x 12 columns]

Cluster 3

In [59]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             2.0        Baseball Field           Yoga Studio   
101   Etobicoke             2.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57     Falafel Restaurant  Ethiopian Restaurant     Electronics Store   
101    Falafel Restaurant  Ethiopian Restaurant     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue  
57                Dog Run    Distribution Center  
101               Dog Run    Distribution Center

Cluster 4

In [60]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             3.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

Cluster 5

In [61]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0   Filipino Restaurant           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11             Drugstore            Donut Shop               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
11   Distribution Center         Discount Store

We have successfully cluster Toronto neighbourhood based on venue categories!